File: nexus-lite/nexus-server/src/request_id.rs
use axum::{
    body::Body,
    http::{HeaderName, HeaderValue, Request},
    middleware::Next,
    response::Response,
};
use uuid::Uuid;

pub const REQUEST_ID_HEADER: HeaderName = HeaderName::from_static("x-request-id");

fn parse_request_id(request: &Request<Body>) -> Option<String> {
    request
        .headers()
        .get(&REQUEST_ID_HEADER)
        .or_else(|| request.headers().get("X-Request-ID"))
        .or_else(|| request.headers().get("X-Request-Id"))
        .and_then(|v| v.to_str().ok())
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
}

pub fn get_or_create_request_id(request: &Request<Body>) -> String {
    parse_request_id(request).unwrap_or_else(|| Uuid::new_v4().to_string())
}

pub async fn ensure_request_id(mut request: Request<Body>, next: Next) -> Response {
    let request_id = get_or_create_request_id(&request);

    // Ensure request header contains the id for downstream services/logging.
    if let Ok(v) = HeaderValue::from_str(&request_id) {
        request.headers_mut().insert(REQUEST_ID_HEADER.clone(), v);
    }

    // Also store it in request extensions for handlers/middleware.
    request.extensions_mut().insert(request_id.clone());

    let mut response = next.run(request).await;
    if response.headers().get(&REQUEST_ID_HEADER).is_none() {
        if let Ok(v) = HeaderValue::from_str(&request_id) {
            response.headers_mut().insert(REQUEST_ID_HEADER.clone(), v);
        }
    }

    response
}

File: edge/shared/request-id.ts
export const REQUEST_ID_HEADER = 'X-Request-ID'

function readRequestId(headers: Headers): string | null {
  return (
    headers.get(REQUEST_ID_HEADER) ||
    headers.get('x-request-id') ||
    headers.get('X-Request-Id')
  )
}

export function getRequestId(request: Request): string {
  return readRequestId(request.headers) || crypto.randomUUID()
}

export function ensureRequestId(request: Request): { request: Request; requestId: string } {
  const requestId = getRequestId(request)
  const current = readRequestId(request.headers)
  if (current === requestId) {
    return { request, requestId }
  }

  const headers = new Headers(request.headers)
  headers.set(REQUEST_ID_HEADER, requestId)
  return {
    request: new Request(request, { headers }),
    requestId,
  }
}
